src/sage/matrix/matrix_polynomial_dense.pyx

```diff
@@ -4079,3 +4084,462 @@ cdef class Matrix_polynomial_dense(Matrix_generic_dense):
+            sage: ccomp.augment(mat).smith_form(transformation=False)
+            [    1     0     0     0]
+            [    0     1     0     0]
+            [    0     0 x + 3     0]
+
+        TESTS:
+
+        Corner cases are handled correctly::
+
+            sage: matrix(ring, 0, 0).basis_completion()
+            []
+            sage: matrix(ring, 0, 0).basis_completion(row_wise=False)
+            []
+            sage: matrix(ring, 0, 0).basis_completion(algorithm="smith")
+            []
+            sage: matrix(ring, 0, 0).basis_completion(row_wise=False,algorithm="smith")
+            []
+            sage: matrix(ring, 0, 2).basis_completion()
+            [1 0]
+            [0 1]
+            sage: matrix(ring, 0, 2).basis_completion(row_wise=False)
+            []
+            sage: matrix(ring, 0, 2).basis_completion(algorithm="smith")
+            [1 0]
+            [0 1]
+            sage: matrix(ring, 0, 2).basis_completion(row_wise=False,algorithm="smith")
+            []
+            sage: matrix(ring, 2, 0).basis_completion()
+            []
+            sage: matrix(ring, 2, 0).basis_completion(row_wise=False)
+            [1 0]
+            [0 1]
+            sage: matrix(ring, 2, 0).basis_completion(algorithm="smith")
+            []
+            sage: matrix(ring, 2, 0).basis_completion(row_wise=False,algorithm="smith")
+            [1 0]
+            [0 1]
+        """
+        if algorithm == "approximant" :
+            if row_wise :
+                return self._basis_completion_via_reversed_approx()
+            else :
+                Ctrsp = self.transpose()._basis_completion_via_reversed_approx()
+                return Ctrsp.transpose()
+
+        elif algorithm == "smith" :
+            ring = self.base_ring()
+            m = self.nrows()
+            n = self.ncols()
+
+            # Smith form, S == U * self * V
+            S, U, V = self.smith_form(transformation=True)
+
+            # S is m x n with rk = rank(self) first diagonal entries nonzero, find rk
+            rk = 0
+            while rk < min(m, n) and not S[rk,rk].is_zero():
+                rk += 1
+
+            # case: matrix is zero (including if m == 0 || n == 0)
+            from sage.matrix.constructor import matrix
+            if rk == 0 :
+                if row_wise :
+                    return matrix.identity(ring, n)
+                else :
+                    return matrix.identity(ring, m)
+
+            # now, matrix is nonzero (and nonempty)
+            if row_wise :
+                C = matrix.identity(ring, n)[rk:,:]
+                VV = V.inverse_of_unit()
+                return C * VV
+            else :
+                C = matrix.identity(ring, m)[:,rk:]
+                UU = U.inverse_of_unit()
+                return UU * C
+
+        else :
+            raise ValueError("algorithm must be one of \"approximant\" or \"smith\".")
+
+
+    def is_basis_completion(self, mat, row_wise=True):
```

Comment:
OK for keeping this method but I cannot really see a use case.
Could you please explain a bit why it is important?

Reply:
(To just add on that, if method is useful internally, name it `_is_basis_completion`.

Reply:
Actually, I wrote this function for testing purposes while writing the ``basis_completion`` method. I agree it should be ``_is_basis_completion`` (I have just done this and pushed), or even removed (to be discussed). I admit that I have no clear use case in mind at the moment, but it is unclear to me that it is useless either: there may be contexts where one may end up one way or another with a candidate completion, and would like to know if it is indeed a completion or not; since completions are not unique, it is not trivial to use ``basis_completion`` to check this. Now that it is there, and tested, what would be reasons for discarding it?